nifi-commons/nifi-site-to-site-client/src/main/java/org/apache/nifi/remote/client/PeerSelector.java

```diff
@@ -16,353 +16,559 @@
+        } catch (final IOException e) {
+            error(logger, eventReporter, "Failed to persist list of peers due to {}; if restarted" +
+                    " and the nodes specified at the remote instance are down," +
+                    " may be unable to transfer data until communications with those nodes are restored", e.toString());
+            logger.error("", e);
+        }
+    }
 
-        logger.debug("Fetching remote peer statuses from: {}", peersToRequestClusterInfoFrom);
-        Exception lastFailure = null;
-        for (final PeerDescription peerDescription : peersToRequestClusterInfoFrom) {
-            try {
-                final Set<PeerStatus> statuses = peerStatusProvider.fetchRemotePeerStatuses(peerDescription);
-                lastFetchedQueryablePeers = statuses.stream()
-                        .filter(p -> p.isQueryForPeers())
-                        .collect(Collectors.toSet());
+    /**
+     * Refreshes the list of S2S peers that flowfiles can be sent to or received from. Uses the stateful
+     * cache to reduce network overhead.
+     */
+    private void refreshPeerStatusCache() {
+        try {
+            // Splitting enumeration and querying into separate methods allows better testing and composition
+            final Set<PeerDescription> peersToQuery = getPeersToQuery();
+            final Set<PeerStatus> statuses = fetchRemotePeerStatuses(peersToQuery);
 
-                return statuses;
-            } catch (final Exception e) {
-                logger.warn("Could not communicate with {}:{} to determine which nodes exist in the remote NiFi cluster, due to {}",
-                        peerDescription.getHostname(), peerDescription.getPort(), e.toString());
-                lastFailure = e;
+            if (statuses.isEmpty()) {
+                logger.info("No peers were retrieved from the remote group {}", peersToQuery.stream().map(p -> p.getHostname() + ":" + p.getPort()).collect(Collectors.joining(",")));
             }
-        }
 
-        final IOException ioe = new IOException("Unable to communicate with remote NiFi cluster in order to determine which nodes exist in the remote cluster");
-        if (lastFailure != null) {
-            ioe.addSuppressed(lastFailure);
+            // Persist the fetched peer statuses
+            PeerStatusCache peerStatusCache = new PeerStatusCache(statuses, System.currentTimeMillis(), peerStatusProvider.getTransportProtocol());
+            persistPeerStatuses(peerStatusCache);
+            logger.info("Successfully refreshed peer status cache; remote group consists of {} peers", statuses.size());
+        } catch (Exception e) {
+            warn(logger, eventReporter, "Unable to refresh remote group peers due to: {}", e.getMessage());
+            if (logger.isDebugEnabled() && e.getCause() != null) {
+                Throwable cause = e.getCause();
+                while ((cause = cause.getCause()) != null) {
```

Comment:
The reason this was introduced was because the exceptions thrown here are often very long (hundreds of lines of stacktrace) and occur very often (if the TLS config is incorrect, this can be thrown many times per second, making the log difficult/impossible to read). I've simplified this to just print the stacktrace if debug is enabled, but I think there is still room for improvement down the line.